diff --git a/malloc.c b/malloc.c
--- a/malloc.c
+++ b/malloc.c
@@ -41,8 +41,21 @@ struct block_meta *find_free_block(struct block_meta **last, size_t size) {
   return current;
 }
 
+#define ALIGN 4
+// para subirlo a 8 necesito saber que META_SIZE % 8
+size_t round_up_4_multiple(size_t size){
+  size_t res = (size/ ALIGN)*ALIGN;
+  if (res<size)
+    res +=ALIGN;
+  return res;
+      /* assert((res/ALIGN)*ALIGN - x <ALIGN); */
+}
+
 struct block_meta *request_space(struct block_meta* last, size_t size) {
   struct block_meta *block;
+  // supossitions: global_size and META_SIZE are also aligned
+  // TODO: assert that
+  size = round_up_4_multiple(size);
   block = sbrk(0); //pa que hacer esto si ya lo obtienes en la siguiente?
   void *request = sbrk(size + META_SIZE);
   assert((void*)block == request); // Not thread safe.
